libass/ass_font.c

```diff
@@ -35,6 +42,197 @@
+        codepage = 932;
+        break;
+    case FT_ENCODING_MS_GB2312:
+        codepage = 936;
+        break;
+    case FT_ENCODING_MS_BIG5:
+        codepage = 950;
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        codepage = 949;
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        codepage = 1361;
+        break;
+    default:
+        return codepoint;
+    }
+
+    // do convert
+    int output_size = WideCharToMultiByte(
+            codepage,
+            0,
+            input_buffer,
+            inbuf_size,
+            output_buffer,
+            4,
+            NULL,
+            &conversion_fail);
+    if (output_size == 0 || conversion_fail) {
+        return codepoint;
+    }
+
+    result = 0;
+    for (size_t i = 0; i < output_size; ++i) {
+        result <<= 8;
+        result |= (uint8_t)output_buffer[i];
+    }
+
+    return result;
+}
+#elif defined(CONFIG_ICONV)
+static const char* SJIS_ENCODING[] = {
+    "CP932",
+    "SHIFT_JIS",
+    "SHIFTJIS",
+    "SJIS",
+    NULL
+};
+
+static const char* GB2312_ENCODING[] = {
+    "CP936",
+    "GB2312",
+    "GB18030",
+    "GBK",
+    NULL
+};
+
+static const char* BIG5_ENCODING[] = {
+    "CP950",
+    "BIG5",
+    NULL
+};
+
+static const char* WANSUNG_ENCODING[] = {
+    "CP949",
+    NULL
+};
+
+static const char* JOHAB_ENCODING[] = {
+    "CP1361",
+    "JOHAB",
+    NULL
+};
+
+static const char* UCS4_ENCODING[] = {
+    "UCS-4LE",
+    "UCS4LE",
+    "UTF32LE",
+    "UTF-32LE",
+    NULL
+};
+
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    uint32_t result = codepoint;
+    char input_buffer[4];
+    char output_buffer[4];
+    char* inbuf = (char*)&input_buffer;
+    char* outbuf = (char*)&output_buffer;
+    size_t inbuf_size = sizeof(input_buffer);
+    size_t outbuf_size = sizeof(output_buffer);
+    const char** encoding_str;
+    const char** ucs4_str = UCS4_ENCODING;
+
+    // map freetype encoding to iconv encoding
+    switch (encoding) {
+    case FT_ENCODING_MS_SJIS:
+        encoding_str = SJIS_ENCODING;
+        break;
+    case FT_ENCODING_MS_GB2312:
+        encoding_str = GB2312_ENCODING;
+        break;
+    case FT_ENCODING_MS_BIG5:
+        encoding_str = BIG5_ENCODING;
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        encoding_str = WANSUNG_ENCODING;
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        encoding_str = JOHAB_ENCODING;
+        break;
+    default:
+        return codepoint;
+    }
+
+    // convert input codepoint to little endian uint32_t bytearray
+    for (int i = 0; i < 4; ++i) {
+        input_buffer[i] = result & 0xFF;
+        result >>= 8;
+    }
+    result = codepoint;
+
+    // open iconv context
+    iconv_t cd = (iconv_t)-1;
+    while (*ucs4_str) {
+        while (*encoding_str) {
+            cd = iconv_open(*encoding_str, *ucs4_str);
+            if (cd != (iconv_t)-1) break;
+            ++encoding_str;
+        }
+        if (cd != (iconv_t)-1) break;
```

Comment:
Style Nit: iirc we usually a space after a cast, so `(icont_t) -1`. Applies to all casts in this patch.

Reply:
Adding a space after a cast looks weird. And existing code doesn't have a space after cast, too.

Reply:
Which variant looks “weird” is subjective. The existing code is admittedly not 100%-consistent, but in `ass_font.c` alone there are 7 instances of casts with a space but only 3 *(all inside one function) without a space. Over the whole codebase the lead of the space-variant is only increasing and e.g. in `ass_rasterizer.c` all casts use a space.
When new code is added we should increase consistency, not make it worse.

Reply:
OK I will do it